matrix/LeastSquaresSolver.hpp

```diff
@@ -0,0 +1,146 @@
+/**
+ * @file LeastSquaresSolver.hpp
+ *
+ * Least Squares Solver using QR householder decomposition.
+ * It calculates x for Ax = b.
+ * A = Q*R
+ * where R is an upper triangular matrix.
+ *
+ * R*x = Q^T*b
+ * This is efficiently solved for x because of the upper triangular property of R.
+ *
+ * @author Bart Slinger <bartslinger@gmail.com>
+ */
+
+#pragma once
+
+#include "math.hpp"
+
+namespace matrix {
+
+template<typename Type, size_t M, size_t N>
+class LeastSquaresSolver
+{
+public:
+
+    /**
+     * @brief Class calculates QR decomposition which can be used for linear
+     * least squares
+     * @param A Matrix of size MxN
+     *
+     * Initialize the class with a MxN matrix. The constructor starts the
+     * QR decomposition. This class does not check the rank of the matrix.
+     * The user needs to make sure that rank(A) = N and N >= M.
+     */
+    LeastSquaresSolver(Matrix<Type, M, N> A)
+    {
+        if (M < N) {
+            return;
+        }
+        // Copy contentents of matrix A
+        _A = A;
+
+        for (size_t j = 0; j < N; j++) {
+            float normx = 0.0f;
```

Comment:
Please use `Type` everywhere instead of `float` and for constants, use `(Type)0.` (without the `f`)

Reply:
Sure about (Type)0. ? -Werror=old-style-cast is enabled

Reply:
Ok, so in that case either `Type(0)` or `static_cast`. I see both being used.